pkg/provider/beyondtrust/provider.go

```diff
@@ -0,0 +1,352 @@
+	return false, fmt.Errorf(errNotImplemented)
+}
+
+// NewClient this is where we initialize the SecretClient and return it for the controller to use.
+func (p *Provider) NewClient(ctx context.Context, store esv1beta1.GenericStore, kube client.Client, namespace string) (esv1beta1.SecretsClient, error) {
+	config := store.GetSpec().Provider.Beyondtrust
+	logger := logging.NewLogrLogger(&ESOLogger)
+	apiURL := config.APIURL
+	certificate := ""
+	certificateKey := ""
+	clientTimeOutInSeconds := 45
+	retryMaxElapsedTimeMinutes := 15
+	separator := "/"
+
+	if config.Separator != "" {
+		separator = config.Separator
+	}
+
+	if config.Clienttimeoutseconds != 0 {
+		clientTimeOutInSeconds = config.Clienttimeoutseconds
+	}
+
+	backoffDefinition := backoff.NewExponentialBackOff()
+	backoffDefinition.InitialInterval = 1 * time.Second
+	backoffDefinition.MaxElapsedTime = time.Duration(retryMaxElapsedTimeMinutes) * time.Second
+	backoffDefinition.RandomizationFactor = 0.5
+
+	clientID, err := loadConfigSecret(ctx, config.Clientid, kube, namespace)
+	if err != nil {
+		return nil, fmt.Errorf("error: %w", err)
+	}
+
+	clientSecret, err := loadConfigSecret(ctx, config.Clientsecret, kube, namespace)
+	if err != nil {
+		return nil, fmt.Errorf("error: %w", err)
+	}
+
+	if config.Certificate != nil && config.Certificatekey != nil {
+		loadedCertificate, err := loadConfigSecret(ctx, config.Certificate, kube, namespace)
+		if err != nil {
+			return nil, fmt.Errorf("error: %w", err)
+		}
+
+		certificate = loadedCertificate
+
+		loadedCertificateKey, err := loadConfigSecret(ctx, config.Certificatekey, kube, namespace)
+		if err != nil {
+			return nil, fmt.Errorf("error: %w", err)
+		}
+
+		certificateKey = loadedCertificateKey
+	}
+
+	// Create an instance of ValidationParams
+	params := utils.ValidationParams{
+		ClientID:                   clientID,
+		ClientSecret:               clientSecret,
+		ApiUrl:                     &apiURL,
+		ClientTimeOutInSeconds:     clientTimeOutInSeconds,
+		Separator:                  &separator,
+		VerifyCa:                   config.VerifyCA,
+		Logger:                     logger,
+		Certificate:                certificate,
+		CertificateKey:             certificateKey,
+		RetryMaxElapsedTimeMinutes: &retryMaxElapsedTimeMinutes,
+		MaxFileSecretSizeBytes:     &maxFileSecretSizeBytes,
+	}
+
+	errorsInInputs := utils.ValidateInputs(params)
+
+	if errorsInInputs != nil {
+		return nil, fmt.Errorf("error: %w", errorsInInputs)
+	}
+
+	// creating a http client
+	httpClientObj, err := utils.GetHttpClient(clientTimeOutInSeconds, config.VerifyCA, certificate, certificateKey, logger)
+
+	if err != nil {
+		return nil, fmt.Errorf("error: %w", err)
+	}
+
+	// instantiating authenticate obj, injecting httpClient object
+	authenticate, _ := auth.Authenticate(*httpClientObj, backoffDefinition, apiURL, clientID, clientSecret, logger, retryMaxElapsedTimeMinutes)
```

Comment:
is this call going to lock the thread while `backoffDefinition` is kicked in? Can you please explain how does this work?

Reply:
Thread is not going to be locked.

_Package backoff implements backoff algorithms for retrying operations.

Use Retry function for retrying operations that may fail. If Retry does not meet your needs, copy/paste the function into your project and modify as you wish._

https://pkg.go.dev/github.com/cenkalti/backoff/v4

it is just for retrying when we have an error from our API